### ISSUE TITLE
Parsing error not caught

### ISSUE THREAD

**Issue by ameenross** [php-integrator/core#82]
So I did some debugging with a project where indexing would consistently stop around 44% progress.

```
The PHP server has something to say: Error while indexing /home/wijnandross/dev/nebula/vendor/mustache/mustache/bin/create_example.php!
PhpParser\Error: Could not parse the code, even after attempting corrections on unknown line in /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/src/Parsing/PartialParser.php:224
Stack trace:
#0 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/src/Indexing/FileIndexer.php(489): PhpIntegrator\Parsing\PartialParser->parse('<?php <<<USAGE\n...')
#1 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/src/Indexing/FileIndexer.php(227): PhpIntegrator\Indexing\FileIndexer->indexConstant(Array, '/home/wijnandro...', '2482', NULL, Object(PhpIntegrator\Analysis\Typing\Resolving\FileTypeResolver))
#2 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/src/Indexing/FileIndexer.php(153): PhpIntegrator\Indexing\FileIndexer->indexVisitorResults('/home/wijnandro...', '2482', Object(PhpIntegrator\Analysis\Visiting\OutlineFetchingVisitor), Object(PhpIntegrator\Analysis\Visiting\UseStatementFetchingVisitor))
#3 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/src/Indexing/ProjectIndexer.php(183): PhpIntegrator\Indexing\FileIndexer->index('/home/wijnandro...', '#!/usr/bin/env ...')
#4 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/src/Indexing/Indexer.php(87): PhpIntegrator\Indexing\ProjectIndexer->index(Array, Array, Array, Array)
#5 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/src/UserInterface/Command/ReindexCommand.php(53): PhpIntegrator\Indexing\Indexer->reindex(Array, false, false, true, Array, Array)
#6 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/src/UserInterface/JsonRpcApplication.php(170): PhpIntegrator\UserInterface\Command\ReindexCommand->execute(Object(ArrayObject))
#7 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/src/UserInterface/JsonRpcApplication.php(117): PhpIntegrator\UserInterface\JsonRpcApplication->handleRequest(Object(PhpIntegrator\Sockets\JsonRpcRequest), Object(PhpIntegrator\Sockets\JsonRpcConnectionHandler))
#8 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/src/Sockets/JsonRpcConnectionHandler.php(175): PhpIntegrator\UserInterface\JsonRpcApplication->handle(Object(PhpIntegrator\Sockets\JsonRpcRequest), Object(PhpIntegrator\Sockets\JsonRpcConnectionHandler))
#9 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/src/Sockets/JsonRpcConnectionHandler.php(147): PhpIntegrator\Sockets\JsonRpcConnectionHandler->getJsonRpcResponseForJsonRpcRequest(Object(PhpIntegrator\Sockets\JsonRpcRequest))
#10 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/src/Sockets/JsonRpcConnectionHandler.php(164): PhpIntegrator\Sockets\JsonRpcConnectionHandler->processData('{"jsonrpc":2,"i...')
#11 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/src/Sockets/JsonRpcConnectionHandler.php(164): PhpIntegrator\Sockets\JsonRpcConnectionHandler->processData('{"jsonrpc":2,"i...')
#12 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/src/Sockets/JsonRpcConnectionHandler.php(79): PhpIntegrator\Sockets\JsonRpcConnectionHandler->processData('\r\n{"jsonrpc":2,...')
#13 [internal function]: PhpIntegrator\Sockets\JsonRpcConnectionHandler->onDataReceived('Content-Length:...', Object(React\Socket\Connection))
#14 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/vendor/evenement/evenement/src/Evenement/EventEmitterTrait.php(64): call_user_func_array(Array, Array)
#15 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/vendor/react/stream/src/Stream.php(173): Evenement\EventEmitter->emit('data', Array)
#16 [internal function]: React\Stream\Stream->handleData(Resource id #49, Object(React\EventLoop\StreamSelectLoop))
#17 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/vendor/react/event-loop/src/StreamSelectLoop.php(232): call_user_func(Array, Resource id #49, Object(React\EventLoop\StreamSelectLoop))
#18 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/vendor/react/event-loop/src/StreamSelectLoop.php(201): React\EventLoop\StreamSelectLoop->waitForStreamActivity(NULL)
#19 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/src/UserInterface/JsonRpcApplication.php(71): React\EventLoop\StreamSelectLoop->run()
#20 /home/wijnandross/.atom/packages/php-integrator-base/core/2.1.2/src/Main.php(70): PhpIntegrator\UserInterface\JsonRpcApplication->run()
#21 {main}
```

It's a file with some rather exotic code, which I can imagine not parsing correctly. However, the parse exception is not caught. When I added the catch block (with the debug output) indexing completed 100%. Here's the file, just in case you're interested.

```php
#!/usr/bin/env php
<?php

/**
 * A commandline script to create an example and the needed files:
 *
 *     $ bin/create_example.php my_new_example
 *
 * ... and the folder my_new_example will be created in the examples/ folder containing 3 files:
 *
 *     my_new_example/my_new_example.mustache
 *     my_new_example/my_new_example.txt
 *     my_new_example/MyNewExample.php
 */

// some constants
define('USAGE', <<<USAGE
USAGE: {$argv[0]} example_name

This creates a new example and the corresponding files in the examples/ directory

USAGE
);

define('EXAMPLE_PATH', realpath(dirname(__FILE__) . '/../test/fixtures/examples'));


/**
 * transform a string to lowercase using underlines.
 * Examples:
 * String -> string
 * AString -> a_string
 * SomeStrings -> some_strings
 * AStringMore -> a_string_more
 *
 * @param string $name
 * @access public
 * @return string
 */
function getLowerCaseName($name) {
    return preg_replace_callback("/([A-Z])/", create_function (
        '$match',
        'return "_" . strtolower($match[1]);'
    ), lcfirst($name));
}

/**
 * transform a string to Uppercase (camelcase)
 * Examples
 * string -> String
 * a_string -> AString
 * some_strings -> SomeStrings
 * a_string_more -> AStringMore -> a_string_more
 *
 * @param string $name
 * @access public
 * @return string
 */
function getUpperCaseName($name) {
    return preg_replace_callback("/_([a-z])/", create_function (
        '$match',
        'return strtoupper($match{1});'
    ), ucfirst($name));
}


/**
 * return the given value and echo it out appending "\n"
 *
 * @param mixed $value
 * @access public
 * @return mixed
 */
function out($value) {
    echo $value . "\n";
    return $value;
}

/**
 * create Path for certain files in an example
 * returns the directory name if only $directory is given.
 * if an extension is given a complete filename is returned.
 * the returned filename will be echoed out
 *
 * @param string $directory directory without / at the end
 * @param string $filename filename without path and extension
 * @param string $extension extension of the file without "."
 * @access public
 * @return string
 */
function buildPath($directory, $filename = null,  $extension = null) {
    return out(EXAMPLE_PATH . '/' . $directory.
                    ($extension !== null && $filename !== null ? '/' . $filename. "." . $extension : ""));
}

/**
 * creates the directory for the example
 * the script die()'s if mkdir() fails
 *
 * @param string $directory
 * @access public
 * @return void
 */
function createDirectory($directory) {
    if(!@mkdir(buildPath($directory))) {
        die("FAILED to create directory\n");
    }
}

/**
 * create a file for the example with the given $content
 * the script die()'s if fopen() fails
 *
 * @param string $directory directory without / at the end
 * @param string $filename filename without path and extension
 * @param string $extension extension of the file without "."
 * @param string $content the content of the file
 * @access public
 * @return void
 */
function createFile($directory, $filename, $extension, $content = "") {
    $handle = @fopen(buildPath($directory, $filename, $extension), "w");
    if($handle) {
        fwrite($handle, $content);
        fclose($handle);
    } else {
        die("FAILED to create file\n");
    }
}


/**
 * routine to create the example directory and 3 files
 *
 * if the $example_name is "SomeThing" the following files will be created
 * examples/some_thing
 * examples/some_thing/some_thing.mustache
 * examples/some_thing/some_thing.txt
 * examples/some_thing/SomeThing.php
 *
 * @param mixed $example_name
 * @access public
 * @return void
 */
function main($example_name) {
    $lowercase = getLowerCaseName($example_name);
    $uppercase = getUpperCaseName($example_name);
    createDirectory($lowercase);
    createFile($lowercase, $lowercase, "mustache");
    createFile($lowercase, $lowercase, "txt");
    createFile($lowercase, $uppercase, "php", <<<CONTENT
<?php

class {$uppercase} {

}

CONTENT
    );
}

// check if enougth arguments are given
if(count($argv) > 1) {
    // get the name of the example
    $example_name = $argv[1];

    main($example_name);

} else {
    echo USAGE;
}

```

**Comment by Gert-dev** [comment]
GitHub autoclosed this, but I've released 2.1.3 along with an update of the base package that will install the new core.

It was the define containing the heredoc, which was a corner case that apparently didn't get parsed correctly, causing indexing of the entire file to fail.

I've also changed the exception handling so that parsing of valid default constant/property values shows up as an error notification inside Atom. This way I can hopefully catch all of the remaining corner cases (if there are any) instead of them being silenced.
